src/lib/tasks/Render.ts

```diff
@@ -89,10 +94,30 @@ export class RenderTask extends Task<RenderTaskParams> {
 
     // --- Basic checks passed we wait a bit more to page to render
     try {
-      // Computing maxWait minus what we already consumed
+      const timeBudget = this.timeBudget.get();
+      const startWaitTime = Date.now();
+
       await this.page.ref?.waitForLoadState('networkidle', {
-        timeout: this.timeBudget.get(),
+        timeout: timeBudget,
       });
+      // waitForLoadState('networkidle') can be flaky and return too soon:
+      // https://github.com/microsoft/playwright/issues/4664#issuecomment-742691215
+      // https://github.com/microsoft/playwright/issues/2515#issuecomment-724163391
+      // So if we still have pending requests, we manually wait.
+      while (
+        this.page.pendingRequests > 0 &&
+        Date.now() - startWaitTime < timeBudget
+      ) {
+        log.debug(
+          { pageUrl: this.page.ref?.url() },
+          `Waiting for ${
+            this.page.pendingRequests
+          } requests to complete... WaitTime:${
+            Date.now() - startWaitTime
+          }, timeBudget: ${timeBudget}`
+        );
+        await setTimeout(1000);
+      }
```

Comment:
do we have a timeout for this ?

Reply:
So the `timeBudget` is always set with a max: https://github.com/algolia/renderscript/blob/master/src/lib/tasks/Task.ts#L68, and the max by default is 20s: https://github.com/algolia/renderscript/blob/master/src/lib/constants.ts#L27
On the Crawler side, we also don't allow to configure it with more than 20s